node/node-core/src/eth/transaction/mod.rs

```diff
@@ -0,0 +1,918 @@
+        &self.nonce
+    }
+
+    pub fn hash(&self) -> H256 {
+        let encoded = rlp::encode(self);
+        let mut out = vec![0; 1 + encoded.len()];
+        out[0] = 2;
+        out[1..].copy_from_slice(&encoded);
+        H256::from_slice(keccak256(&out).as_slice())
+    }
+
+    /// Recovers the Ethereum address which was used to sign the transaction.
+    pub fn recover(&self) -> Result<Address, SignatureError> {
+        let mut sig = [0u8; 65];
+        sig[0..32].copy_from_slice(&self.r[..]);
+        sig[32..64].copy_from_slice(&self.s[..]);
+        sig[64] = self.odd_y_parity as u8;
+        let signature = Signature::try_from(&sig[..])?;
+        signature.recover(EIP1559TransactionRequest::from(self.clone()).hash())
+    }
+}
+
+impl Encodable for EIP1559Transaction {
+    fn rlp_append(&self, s: &mut RlpStream) {
+        s.begin_list(12);
+        s.append(&self.chain_id);
+        s.append(&self.nonce);
+        s.append(&self.max_priority_fee_per_gas);
+        s.append(&self.max_fee_per_gas);
+        s.append(&self.gas_limit);
+        s.append(&self.kind);
+        s.append(&self.value);
+        s.append(&self.input.as_ref());
+        s.append(&self.access_list);
+        s.append(&self.odd_y_parity);
+        s.append(&U256::from_big_endian(&self.r[..]));
+        s.append(&U256::from_big_endian(&self.s[..]));
+    }
+}
+
+impl Decodable for EIP1559Transaction {
+    fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+        if rlp.item_count()? != 12 {
+            return Err(DecoderError::RlpIncorrectListLen)
+        }
+
+        Ok(Self {
+            chain_id: rlp.val_at(0)?,
+            nonce: rlp.val_at(1)?,
+            max_priority_fee_per_gas: rlp.val_at(2)?,
+            max_fee_per_gas: rlp.val_at(3)?,
+            gas_limit: rlp.val_at(4)?,
+            kind: rlp.val_at(5)?,
+            value: rlp.val_at(6)?,
+            input: rlp.val_at::<Vec<u8>>(7)?.into(),
+            access_list: rlp.val_at(8)?,
+            odd_y_parity: rlp.val_at(9)?,
+            r: {
+                let mut rarr = [0u8; 32];
+                rlp.val_at::<U256>(10)?.to_big_endian(&mut rarr);
+                H256::from(rarr)
+            },
+            s: {
+                let mut sarr = [0u8; 32];
+                rlp.val_at::<U256>(11)?.to_big_endian(&mut sarr);
+                H256::from(sarr)
+            },
+        })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TransactionEssentials {
+    pub kind: TransactionKind,
+    pub input: Bytes,
+    pub nonce: U256,
+    pub gas_limit: U256,
+    pub gas_price: Option<U256>,
+    pub max_fee_per_gas: Option<U256>,
+    pub max_priority_fee_per_gas: Option<U256>,
+    pub value: U256,
+    pub chain_id: Option<u64>,
+    pub access_list: AccessList,
+}
+
+/// Queued transaction
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PendingTransaction {
+    /// The actual transaction
+    pub transaction: TypedTransaction,
```

Comment:
should we try to use ethers' TypedTransaction request instead and get rid of the above boilerplate/compat types?